sim/xcelium/__init__.py

```diff
@@ -0,0 +1,394 @@
+
+    # Create saif file if specified.
+    if saif_opts["mode"] is not None:
+      f.write(f'{self.generate_saif_tcl_cmd()}\n')
+    
+    # Execute
+    f.write("run \n")
+    
+    # Close databases and dumps properly.
+    f.write("dumpsaif -end \n")
+    f.write("database -close *db \n")
+    f.write("exit")
+    f.close()  
+    return True
+
+  def compile_xrun(self) -> bool:
+    
+    if not os.path.isfile(self.xcelium_bin):
+      self.logger.error(f"Xcelium (xrun) binary not found at {self.xcelium_bin}.")
+      return False
+  
+    if not self.check_input_files(self.xcelium_ext):
+      return False
+
+    # Gather complation-only options
+    compile_opts  = self.get_setting(f"{self.tool_config_prefix}.compile_opts", [])       
+    compile_opts.append("-logfile xrun_compile.log")
+    compile_opts = ('COMPILE', compile_opts)
+    
+    arg_file_path = self.generate_arg_file("xrun_compile.arg", "HAMMER-GEN XRUN COMPILE ARG FILE", [compile_opts])
+    args =[self.xcelium_bin]
+    args.append(f"-compile -f {arg_file_path}")
+    
+    self.run_executable(args, cwd=self.run_dir)
+    HammerVLSILogging.enable_colour = True
+    HammerVLSILogging.enable_tag = True
+    return True
+    
+  def elaborate_xrun(self) -> bool: 
+    sim_opts  = self.extract_sim_opts()[1]
+    elab_opts = self.get_setting(f"{self.tool_config_prefix}.elab_opts", [])
+    elab_opts.append("-logfile xrun_elab.log")
+    elab_opts.append("-glsperf")
+    elab_opts.append("-genafile access.txt")  
+    
+    if self.level.is_gatelevel():
+      elab_opts.extend(self.get_verilog_models())    
+      if sim_opts["timing_annotated"]:
+        self.generate_sdf_cmd_file()
+        elab_opts.append(f"-sdf_cmd_file {self.sdf_cmd_file}")  
+        elab_opts.append("-sdf_verbose")
+        elab_opts.append("-negdelay")
+      else:
+        elab_opts.append("-notimingchecks")
+        elab_opts.append("-delay_mode zero")
+    else:
+      elab_opts.append("-notimingchecks")
+      elab_opts.append("-delay_mode zero")
+    elab_opts = ('ELABORATION', elab_opts)
+    
+    arg_file_path = self.generate_arg_file("xrun_elab.arg", "HAMMER-GEN XRUN ELAB ARG FILE", [elab_opts])
+    args =[self.xcelium_bin]
+    args.append(f"-elaborate -f {arg_file_path}")
+    
+    self.run_executable(args, cwd=self.run_dir)
+    return True
+
+  def sim_xrun(self) -> bool:
+    
+    sim_opts = self.get_setting(f"{self.sim_input_prefix}.options", [])
+    sim_opts_removal  = ["tb_name", "input_files", "incdir"]
+    xrun_opts_removal = ["enhanced_recompile"]
+    sim_opts = ('SIMULATION', sim_opts)
+    
+    arg_file_path = self.generate_arg_file("xrun_sim.arg", "HAMMER-GEN XRUN SIM ARG FILE", [sim_opts],
+                                           sim_opt_removal = sim_opts_removal,
+                                           xrun_opt_removal = xrun_opts_removal)    
+    args =[self.xcelium_bin]
+    args.append(f"-R -f {arg_file_path} -input {self.sim_tcl_file}")
```

Comment:
Is there a use for appending `sim.inputs.execution_flags` or `sim.inputs.options`? Perhaps `sim.inputs.options` could replace `sim.xcelium.elab_opts`?

Reply:
sim.xcelium.compile_opts, sim.xcelium.elab_opts are custom options mirroring the compilation and elaboration steps. The intention is to use these keys to provide cleaner and deeper access into options for each of the steps. More generally, the plugin was structured like this instead of a single xrun "super-command" because accessing individual tool tcl/cmd files, option scopes, saving snapshots/envs become less clear as the project gets larger. I think this will be especially true when AMS support is implemented.

Sim.input.options corresponds to options for xmsim tool in this plugin. Ideally it'd move it under sim.xcelium but since the bulk of sim.input.options that I've seen (open to correction) are specific to the simulation step this is fine.

If elaboration options were included in prior testbenches under sim.input.options, they will need to be updated to be moved under sim.xcelium.elab_opts. However, because elab opts differ from VCS to Xcelium, changes are in order anyway and a bit of minor restructuring could potentially be good.

Reply:
So, I'm trying to advocate here for not moving keys to `sim.xcelium` and keeping keys as globally abstract as possible.

Anyways, yes, I think you've implemented `sim.inputs.options` correctly. `sim.inputs.execution_flags` is actually used for RISC-V program/benchmark simulation (i.e. passing in args to the program, not the simulator). You haven't implemented this anyway yet, so I'm OK with punting it to the next PR.

Given the above I see now that `sim.xcelium.compile_opts` and `sim.xcelium.elab_opts` are useful/fine to keep in here, since there's no straightforward analogue in VCS.

Reply:
Missed out on `sim.inputs.execution_flags` but I can get that on the next pass.